linker/shared/src/main/scala/org/scalajs/linker/frontend/modulesplitter/MaxModuleAnalyzer.scala

```diff
@@ -97,28 +109,121 @@ private object MaxModuleAnalyzer {
       map.get(className)
   }
 
+  /** Tagger performs the actual grouping of classes into modules.
+   *
+   *  To group classes into modules appropriately, we want to know for
+   *  each class, "how" it can be reached. In practice, this means we
+   *  record the path from the original public module and every
+   *  dynamic import hop we made.
+   *
+   *  Of all these paths, we only care about the "simplest" ones. Or
+   *  more formally, the minimum prefixes of all paths. For example,
+   *  if a class is reachable by the following paths:
+   *
+   *  - a -> b
+   *  - a -> b -> c
+   *  - d -> c
+   *  - d
+   *
+   *  We really only care about:
+   *
+   *  - a -> b
+   *  - d
+   *
+   *  Because if we reach the class through a path that goes through
+   *  `c`, it is necessarily already loaded.
+   *
+   *  Once we have obtained this minimal set of paths, we use the last
+   *  elements of the paths to determine the final module
+   *  grouping. This is because these have an actual static dependency
+   *  on the node in question.
+   *
+   *  Merging these tags into a single `ModuleID` is delegated to the
+   *  caller.
+   */
```

Comment:
@sjrd PTAL at this. Is it convincing enough that the algorithm is correct?

Reply:
Yes, exactly. But isn't this exactly the example I'm making here with your `c` being "the class", your `a` being `d` and your `b` being `c`?

From the simplification we can see that `c` (your `b`) does not appear in the final grouping.

I'm not sure I know what to explain better or how based on your comment :-/